torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -1753,22 +1794,74 @@ struct to_ir {
     };
     GatheredTypes gathered(typeParser_);
     gathered.gather(classinfo);
-    auto val = emitExpr(obj);
+    Value* val = emitExpr(obj);
     RefinementSet refinement;
-    if (gathered.types.size() == 1 &&
-        gathered.types.at(0)->isSubtypeOf(val->type()) &&
-        obj.kind() == TK_VAR) {
+
+    TORCH_CHECK(!gathered.types.empty(), "`isinstance` must be used with a "
+                "type or a tuple of types");
+
+    bool all_are_subtypes = std::all_of(gathered.types.begin(), gathered.types.end(),
+                                        [&](const TypePtr t) {
+                                          return t->isSubtypeOf(val->type());
+                                        });
+    if (all_are_subtypes && obj.kind() == TK_VAR) {
       std::string ident = Var(obj).name().name();
-      Refinement isinstance(std::move(ident), gathered.types.at(0));
-      refinement = RefinementSet({isinstance}, {});
+
+      // Get the type to compare against. If we have a tuple of types,
+      // turn the tuple into a Union type to prevent unnecessary extra
+      // logic for single vs. multiple comparisons
+      const TypePtr single_comparison = gathered.types.size() == 1 ?
+                                        gathered.types[0] :
+                                        UnionType::create(gathered.types);
+
+      Refinement isinstance(ident, single_comparison);
+
+      if (val->type()->kind() == UnionType::Kind) {
+        const UnionTypePtr union_type = val->type()->expect<UnionType>();
+
+        std::vector<TypePtr> not_isinstance_types;
+
+        // O(1) lookups of all the types in `gathered.types`
+        std::unordered_set<TypePtr, std::hash<TypePtr>, c10::TypeEqual> dict{
+          gathered.types.begin(), gathered.types.end()};
+
+        for (auto type : union_type->containedTypes()) {
+          // Fast path
+          if (!dict.count(type)) {
+            not_isinstance_types.emplace_back(type);
+            break;
+          }
+          // Check all types in `gathered.types` to see if there's a
+          // common supertype
+          for (const TypePtr comparison_type : gathered.types) {
```

Comment:
im a little confused by this logic, why are we unifying the comparison_type with the type?